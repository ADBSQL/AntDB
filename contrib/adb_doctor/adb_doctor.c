/*--------------------------------------------------------------------------
 * 
 * Copyright (c) 2018-2019, Asiainfo Database Innovation Lab
 * 
 * user interface of control adb_doctor, support function below:
 * select adb_doctor_start,
 * select adb_doctor_stop,
 * select adb_doctor_param.
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"
#include "fmgr.h"
#include "miscadmin.h"
#include "pgstat.h"
#include "storage/dsm.h"
#include "storage/shm_mq.h"
#include "storage/shm_toc.h"
#include "storage/spin.h"
#include "utils/memutils.h"
#include "utils/builtins.h"
#include "utils/snapmgr.h"
#include "executor/spi.h"
#include "access/xact.h"
#include "access/xlog.h"
#include "postmaster/bgworker.h"
#include "adb_doctor.h"
#include "adb_doctor_sql.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(adb_doctor_start);
PG_FUNCTION_INFO_V1(adb_doctor_stop);
PG_FUNCTION_INFO_V1(adb_doctor_param);

void _PG_init(void);

static shm_mq *setupShmMQ(dsm_segment **segp);
static BackgroundWorkerHandle *startupLauncher(dsm_segment *seg);
static bool isLauncherOK(Size len, char *message);
static void waitForLauncherOK(BackgroundWorkerHandle *launcherHandle, shm_mq_handle *inqh);

/**
 * Start all doctor process.
 * Register an "adb doctor launcher" process with "postmaster", which runs as a 
 * background worker. The "adb doctor launcher" will start the doctor process 
 * according to the table MGR_NODE, MGR_HOST in the MGR database.
 */
Datum
	adb_doctor_start(PG_FUNCTION_ARGS)
{
	dsm_segment *seg;
	shm_mq *mq;
	shm_mq_handle *inqh;
	BackgroundWorkerHandle *launcherHandle;
	bool masterMode;

	masterMode = !RecoveryInProgress();

	adbDoctorStopLauncher(false);

	mq = setupShmMQ(&seg);
	launcherHandle = startupLauncher(seg);

	on_dsm_detach(seg, cleanupAdbDoctorBgworker, PointerGetDatum(launcherHandle));

	if (masterMode)
	{
		inqh = shm_mq_attach(mq, seg, launcherHandle);
		waitForLauncherOK(launcherHandle, inqh);
	}

	cancel_on_dsm_detach(seg, cleanupAdbDoctorBgworker, PointerGetDatum(launcherHandle));

	dsm_detach(seg);
	PG_RETURN_VOID();
}

/**
 * Stop all doctor processes.
 */
Datum
	adb_doctor_stop(PG_FUNCTION_ARGS)
{
	adbDoctorStopLauncher(false);
	adbDoctorStopBgworkers(false);
	PG_RETURN_VOID();
}

/**
 * Set configuration variables stored in table adb_doctor_conf.
 * Use like this: select adb_doctor.adb_doctor_param('name', 'value');
 */
Datum
	adb_doctor_param(PG_FUNCTION_ARGS)
{
	text *k_txt = PG_GETARG_TEXT_PP(0);
	text *v_txt = PG_GETARG_TEXT_PP(1);
	char *k;
	char *v;
	int ret;

	k = text_to_cstring(k_txt);
	v = text_to_cstring(v_txt);

	validateAdbDoctorConfElement(k, v);
	ret = SPI_connect();
	if (ret != SPI_OK_CONNECT)
	{
		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
						(errmsg("SPI_connect failed, connect return:%d", ret))));
	}
	SPI_updateAdbDoctorConf(k, v);
	SPI_finish();

	adbDoctorSignalLauncher();

	pfree(k);
	pfree(v);
	PG_RETURN_VOID();
}

static shm_mq *
setupShmMQ(dsm_segment **segp)
{
	Size queue_size = 96;
	shm_toc_estimator e;
	dsm_segment *seg;
	Size segsize;
	shm_toc *toc;
	shm_mq *mq;

	shm_toc_initialize_estimator(&e);
	shm_toc_estimate_chunk(&e, queue_size);
	shm_toc_estimate_keys(&e, 1);
	segsize = shm_toc_estimate(&e);

	seg = dsm_create(segsize, 0);
	*segp = seg;
	toc = shm_toc_create(ADB_DOCTOR_LAUNCHER_MAGIC, dsm_segment_address(seg),
						 segsize);

	mq = shm_mq_create(shm_toc_allocate(toc, (Size)queue_size), (Size)queue_size);
	shm_toc_insert(toc, 0, mq);
	shm_mq_set_receiver(mq, MyProc);
	return mq;
}

static BackgroundWorkerHandle *startupLauncher(dsm_segment *seg)
{
	BackgroundWorker worker;
	BackgroundWorkerHandle *handle;

	/* Configure a worker. */
	memset(&worker, 0, sizeof(BackgroundWorker));
	worker.bgw_flags = BGWORKER_SHMEM_ACCESS |
					   BGWORKER_BACKEND_DATABASE_CONNECTION;
	worker.bgw_start_time = BgWorkerStart_RecoveryFinished;
	worker.bgw_restart_time = 30;
	sprintf(worker.bgw_library_name, ADB_DOCTOR_BGW_LIBRARY_NAME);
	sprintf(worker.bgw_function_name, ADB_DOCTOR_FUNCTION_NAME_LAUNCHER);
	snprintf(worker.bgw_name, BGW_MAXLEN, ADB_DOCTOR_BGW_TYPE_LAUNCHER);
	snprintf(worker.bgw_type, BGW_MAXLEN, ADB_DOCTOR_BGW_TYPE_LAUNCHER);
	worker.bgw_main_arg = UInt32GetDatum(dsm_segment_handle(seg));
	worker.bgw_notify_pid = MyProcPid;

	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
		ereport(ERROR,
				(errcode(ERRCODE_INSUFFICIENT_RESOURCES),
				 errmsg("could not register background process"),
				 errhint("You may need to increase max_worker_processes.")));
	ereport(LOG,
			(errmsg("register adb doctor launcher success")));

	return handle;
}

static bool isLauncherOK(Size len, char *message)
{
	Size expectedLen;

	expectedLen = strlen(ADB_DOCTORS_LAUNCH_OK);
	if (len != expectedLen)
		return false;
	if (strcmp(message, ADB_DOCTORS_LAUNCH_OK) != 0)
		return false;
	return true;
}

static void waitForLauncherOK(BackgroundWorkerHandle *launcherHandle, shm_mq_handle *inqh)
{
	BgwHandleStatus status;
	pid_t pid;
	shm_mq_result res;
	Size len;
	void *message;

	/* wait for postmaster startup the worker, and then we setup next worker */
	status = WaitForBackgroundWorkerStartup(launcherHandle, &pid);
	if (status != BGWH_STARTED)
		ereport(ERROR,
				(errcode(ERRCODE_INSUFFICIENT_RESOURCES),
				 errmsg("could not start background process"),
				 errhint("More details may be available in the server log.")));

	while (true)
	{
		ProcessInterrupts();
		res = shm_mq_receive(inqh, &len, &message, false);
		if (res == SHM_MQ_SUCCESS)
		{
			if (!isLauncherOK(len, message))
			{
				ereport(ERROR,
						(errmsg("launch doctor worker failed"),
						 errdetail("please view server log for detail")));
			}
			break;
		}
		else
		{
			pg_usleep(100000L);
		}
	}
}

void adbDoctorStopLauncher(bool waitForStopped)
{
	TerminateBackgroundWorkerByBgwType(ADB_DOCTOR_BGW_LIBRARY_NAME, ADB_DOCTOR_BGW_TYPE_LAUNCHER, waitForStopped);
}

void adbDoctorStopBgworkers(bool waitForStopped)
{
	TerminateBackgroundWorkerByBgwType(ADB_DOCTOR_BGW_LIBRARY_NAME, ADB_DOCTOR_BGW_TYPE_WORKER, waitForStopped);
}

void adbDoctorSignalLauncher(void)
{
	ReportToBackgroundWorkerByBgwType(ADB_DOCTOR_BGW_LIBRARY_NAME, ADB_DOCTOR_BGW_TYPE_LAUNCHER);
}

void cleanupAdbDoctorBgworker(dsm_segment *seg, Datum arg)
{
	BackgroundWorkerHandle *handle = (BackgroundWorkerHandle *)DatumGetPointer(arg);
	if (handle != NULL)
	{
		TerminateBackgroundWorker(handle);
	}
}

void notifyAdbDoctorRegistrant(void)
{
	PGPROC *registrant = BackendPidGetProc(MyBgworkerEntry->bgw_notify_pid);
	if (registrant == NULL)
	{
		ereport(ERROR,
				(errmsg("registrant backend has exited prematurely")));
		proc_exit(0);
	}
	SetLatch(&registrant->procLatch);
}

void usleepIgnoreSignal(long microsec)
{
	TimestampTz current;
	TimestampTz latest;
	current = GetCurrentTimestamp();
	while (microsec > 0)
	{
		latest = current;
		pg_usleep(microsec);
		current = GetCurrentTimestamp();
		microsec -= (current - latest);
	}
}