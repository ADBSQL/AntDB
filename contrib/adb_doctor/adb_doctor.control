comment = 'cure the adb cluster automatically'
default_version = '1.0'
module_pathname = '$libdir/adb_doctor'
relocatable = false
schema = adb_doctor