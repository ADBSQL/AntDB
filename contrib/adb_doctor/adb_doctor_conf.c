/*--------------------------------------------------------------------------
 *
 * Copyright (c) 2018-2019, Asiainfo Database Innovation Lab
 *
 * -------------------------------------------------------------------------
 */
#include "postgres.h"
#include "storage/dsm.h"
#include "storage/shm_toc.h"
#include "utils/resowner.h"
#include "storage/lwlock.h"
#include "adb_doctor_conf.h"

void checkAdbDoctorConf(AdbDoctorConf *src)
{
	src->forceswitch = LIMIT_VALUE_RANGE(0, 1, src->forceswitch);
	src->switchinterval = LIMIT_VALUE_RANGE(ADB_DOCTOR_CONF_SWITCHINTERVAL_MIN,
											ADB_DOCTOR_CONF_SWITCHINTERVAL_MAX,
											src->switchinterval);
	src->nodedeadline = LIMIT_VALUE_RANGE(ADB_DOCTOR_CONF_NODEDEADLINE_MIN,
										  ADB_DOCTOR_CONF_NODEDEADLINE_MAX,
										  src->nodedeadline);
	src->agentdeadline = LIMIT_VALUE_RANGE(ADB_DOCTOR_CONF_AGENTDEADLINE_MIN,
										   ADB_DOCTOR_CONF_AGENTDEADLINE_MAX,
										   src->agentdeadline);
	src->node_restart_crashed_master = LIMIT_VALUE_RANGE(0, 1,
														 src->node_restart_crashed_master);
	if (src->node_restart_master_timeout_ms < 1)
		ereport(ERROR,
				(errmsg("node_restart_master_timeout_ms must > 0")));
	if (src->node_shutdown_timeout_ms < 1)
		ereport(ERROR,
				(errmsg("node_shutdown_timeout_ms must > 0")));
	if (src->node_connection_error_num_max < 1)
		ereport(ERROR,
				(errmsg("node_connection_error_num_max must > 0")));
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  node_connect_timeout_ms_min,
								  node_connect_timeout_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  node_reconnect_delay_ms_min,
								  node_reconnect_delay_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  node_query_timeout_ms_min,
								  node_query_timeout_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  node_query_interval_ms_min,
								  node_query_interval_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  node_restart_delay_ms_min,
								  node_restart_delay_ms_max);
	if (src->agent_connection_error_num_max < 1)
		ereport(ERROR,
				(errmsg("agent_connection_error_num_max must > 0")));
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  agent_connect_timeout_ms_min,
								  agent_connect_timeout_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  agent_reconnect_delay_ms_min,
								  agent_reconnect_delay_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  agent_heartbeat_timeout_ms_min,
								  agent_heartbeat_timeout_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  agent_heartbeat_interval_ms_min,
								  agent_heartbeat_interval_ms_max);
	CHECK_ADB_DOCTOR_CONF_MIN_MAX(src,
								  agent_restart_delay_ms_min,
								  agent_restart_delay_ms_max);
}

bool equalsAdbDoctorConfIgnoreLock(AdbDoctorConf *conf1,
								   AdbDoctorConf *conf2)
{
	Size addrOffset;
	Size objectSize = sizeof(AdbDoctorConf);
	Size lockSize = sizeof(LWLock);
	Size lockOffset = offsetof(AdbDoctorConf, lock);
	/* if the lock is the first element? */
	if (lockOffset > 0)
	{
		if (memcmp(conf1, conf2, lockOffset) != 0)
			return false;
	}
	addrOffset = lockOffset + lockSize;
	/* if the lock is the last element? */
	if (addrOffset < objectSize)
	{
		if (memcmp(((char *)conf1) + addrOffset,
				   ((char *)conf2) + addrOffset,
				   objectSize - addrOffset) != 0)
			return false;
	}
	return true;
}

AdbDoctorConfShm *setupAdbDoctorConfShm(AdbDoctorConf *conf)
{
	shm_toc_estimator e;
	dsm_segment *seg;
	Size segsize;
	shm_toc *toc;
	int tocKey = 0;
	AdbDoctorConf *confInShm;
	AdbDoctorConfShm *confShm;

	shm_toc_initialize_estimator(&e);
	shm_toc_estimate_chunk(&e, sizeof(AdbDoctorConf));
	shm_toc_estimate_keys(&e, 1);
	segsize = shm_toc_estimate(&e);

	seg = dsm_create(segsize, 0);

	toc = shm_toc_create(ADB_DOCTOR_CONF_SHM_MAGIC, dsm_segment_address(seg),
						 segsize);

	confInShm = shm_toc_allocate(toc, sizeof(AdbDoctorConf));
	memcpy(confInShm, conf, sizeof(AdbDoctorConf));

	LWLockInitialize(&confInShm->lock, LWLockNewTrancheId());
	LWLockRegisterTranche(confInShm->lock.tranche, "adb_doctor_conf");

	shm_toc_insert(toc, tocKey++, confInShm);

	confShm = palloc0(sizeof(AdbDoctorConfShm));
	confShm->seg = seg;
	confShm->toc = toc;
	confShm->confInShm = confInShm;
	return confShm;
}

AdbDoctorConfShm *attachAdbDoctorConfShm(dsm_handle handle, char *name)
{
	dsm_segment *seg;
	shm_toc *toc;
	AdbDoctorConf *confInShm;
	int tocKey = 0;
	AdbDoctorConfShm *confShm;

	CurrentResourceOwner = ResourceOwnerCreate(NULL, name);
	seg = dsm_attach(handle);
	if (seg == NULL)
		ereport(ERROR,
				(errmsg("unable to map common dynamic shared memory segment")));

	toc = shm_toc_attach(ADB_DOCTOR_CONF_SHM_MAGIC, dsm_segment_address(seg));
	if (toc == NULL)
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("bad magic number in dynamic shared memory segment")));

	confInShm = shm_toc_lookup(toc, tocKey++, false);

	confShm = palloc0(sizeof(AdbDoctorConfShm));
	confShm->seg = seg;
	confShm->toc = toc;
	confShm->confInShm = confInShm;
	return confShm;
}

AdbDoctorConf *copyAdbDoctorConfFromShm(AdbDoctorConfShm *shm)
{
	AdbDoctorConf *confInShm;
	AdbDoctorConf *copy;

	Assert(shm);
	confInShm = shm->confInShm;
	Assert(confInShm);

	copy = palloc(sizeof(AdbDoctorConf));
	LWLockAcquire(&confInShm->lock, LW_SHARED);
	memcpy(copy, confInShm, sizeof(AdbDoctorConf));
	LWLockRelease(&confInShm->lock);
	return copy;
}

/**
 * Memory copy bypass the element lock. 
 */
void copyAdbDoctorConfAvoidLock(AdbDoctorConf *dest,
								AdbDoctorConf *src)
{
	Size addrOffset;
	Size objectSize = sizeof(AdbDoctorConf);
	Size lockSize = sizeof(LWLock);
	Size lockOffset = offsetof(AdbDoctorConf, lock);
	/* if the lock is the first element? */
	if (lockOffset > 0)
	{
		/* Copy the data before lock */
		memcpy(dest, src, lockOffset);
	}
	addrOffset = lockOffset + lockSize;
	/* if the lock is the last element? */
	if (addrOffset < objectSize)
	{
		/* Copy the data after lock */
		memcpy(((char *)dest) + addrOffset,
			   ((char *)src) + addrOffset,
			   objectSize - addrOffset);
	}
}

void refreshAdbDoctorConfInShm(AdbDoctorConf *confInLocal,
							   AdbDoctorConfShm *shm)
{
	AdbDoctorConf *confInShm;

	Assert(shm);
	confInShm = shm->confInShm;
	Assert(confInShm);

	LWLockAcquire(&confInShm->lock, LW_EXCLUSIVE);
	copyAdbDoctorConfAvoidLock(confInShm, confInLocal);
	LWLockRelease(&confInShm->lock);
}

void validateAdbDoctorConfEditableEntry(char *k, char *v)
{
	int v_int;
	if (k == NULL || strlen(k) == 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("parameter k:%s must not empty", k)));
	if (v == NULL || strlen(v) == 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("parameter v:%s must not empty", v)));
	v_int = pg_atoi(v, sizeof(int), 0);
	if (pg_strcasecmp(k, ADB_DOCTOR_CONF_KEY_FORCESWITCH) == 0)
	{
		if (v_int < 0 || v_int > 1)
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("forceswitch must between %d and %d",
							0,
							1)));
	}
	else if (pg_strcasecmp(k, ADB_DOCTOR_CONF_KEY_SWITCHINTERVAL) == 0)
	{
		if (v_int < ADB_DOCTOR_CONF_SWITCHINTERVAL_MIN ||
			v_int > ADB_DOCTOR_CONF_SWITCHINTERVAL_MAX)
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("switchinterval must between %d and %d",
							ADB_DOCTOR_CONF_SWITCHINTERVAL_MIN,
							ADB_DOCTOR_CONF_SWITCHINTERVAL_MAX)));
	}
	else if (pg_strcasecmp(k, ADB_DOCTOR_CONF_KEY_NODEDEADLINE) == 0)
	{
		if (v_int < ADB_DOCTOR_CONF_NODEDEADLINE_MIN ||
			v_int > ADB_DOCTOR_CONF_NODEDEADLINE_MAX)
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("nodedeadline must between %d and %d",
							ADB_DOCTOR_CONF_NODEDEADLINE_MIN,
							ADB_DOCTOR_CONF_NODEDEADLINE_MAX)));
	}
	else if (pg_strcasecmp(k, ADB_DOCTOR_CONF_KEY_AGENTDEADLINE) == 0)
	{
		if (v_int < ADB_DOCTOR_CONF_AGENTDEADLINE_MIN ||
			v_int > ADB_DOCTOR_CONF_AGENTDEADLINE_MAX)
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("agentdeadline must between %d and %d",
							ADB_DOCTOR_CONF_AGENTDEADLINE_MIN,
							ADB_DOCTOR_CONF_AGENTDEADLINE_MAX)));
	}
	else
	{
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("unrecognize adb doctor conf key:%s",
						k)));
	}
}