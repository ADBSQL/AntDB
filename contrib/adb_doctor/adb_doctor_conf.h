/*--------------------------------------------------------------------------
 *
 * Copyright (c) 2018-2019, Asiainfo Database Innovation Lab
 *
 * -------------------------------------------------------------------------
 */
#ifndef ADB_DOCTOR_CONF_H
#define ADB_DOCTOR_CONF_H

#include "storage/lwlock.h"
#include "storage/dsm.h"
#include "storage/shm_toc.h"
#include "utils/builtins.h"
#include "adb_doctor_utils.h"

#define ADB_DOCTOR_CONF_SWITCHINTERVAL_MIN 1
#define ADB_DOCTOR_CONF_SWITCHINTERVAL_MAX 7200
#define ADB_DOCTOR_CONF_NODEDEADLINE_MIN 1
#define ADB_DOCTOR_CONF_NODEDEADLINE_MAX 7200
#define ADB_DOCTOR_CONF_AGENTDEADLINE_MIN 1
#define ADB_DOCTOR_CONF_AGENTDEADLINE_MAX 7200

#define ADB_DOCTOR_CONF_SHM_MAGIC 0x79fb2449

#define ADB_DOCTOR_SCHEMA "adb_doctor"
#define ADB_DOCTOR_CONF_RELNAME "adb_doctor_conf"
#define ADB_DOCTOR_CONF_ATTR_KEY "k"
#define ADB_DOCTOR_CONF_ATTR_VALUE "v"
#define ADB_DOCTOR_CONF_ATTR_COMMENT "comment"
#define ADB_DOCTOR_CONF_KEY_FORCESWITCH "forceswitch"
#define ADB_DOCTOR_CONF_KEY_SWITCHINTERVAL "switchinterval"
#define ADB_DOCTOR_CONF_KEY_NODEDEADLINE "nodedeadline"
#define ADB_DOCTOR_CONF_KEY_AGENTDEADLINE "agentdeadline"

#define CHECK_ADB_DOCTOR_CONF_MIN_MAX(ptr, minMember, maxMember) \
	do                                                           \
	{                                                            \
		if (ptr->minMember < 1)                                  \
			ereport(ERROR,                                       \
					(errmsg(#minMember " must > 0")));           \
		if (ptr->maxMember < 1)                                  \
			ereport(ERROR,                                       \
					(errmsg(#maxMember " must > 0")));           \
		if (ptr->minMember > ptr->maxMember)                     \
			ereport(ERROR,                                       \
					(errmsg(#maxMember " must > " #minMember))); \
	} while (0)

/* AdbDoctorConf elements all in one */
typedef struct AdbDoctorConf
{
	LWLock lock;
	/* Below three elements are editable */
	int forceswitch;
	int switchinterval;
	int nodedeadline;
	int agentdeadline;
	/* The elements below are not editable, keep these member names 
	 * the same as the k field values in the table adb_doctor_conf */
	int node_restart_crashed_master;
	int node_restart_master_timeout_ms;
	int node_shutdown_timeout_ms;
	int node_connection_error_num_max;
	int node_connect_timeout_ms_min;
	int node_connect_timeout_ms_max;
	int node_reconnect_delay_ms_min;
	int node_reconnect_delay_ms_max;
	int node_query_timeout_ms_min;
	int node_query_timeout_ms_max;
	int node_query_interval_ms_min;
	int node_query_interval_ms_max;
	int node_restart_delay_ms_min;
	int node_restart_delay_ms_max;
	int agent_connection_error_num_max;
	int agent_connect_timeout_ms_min;
	int agent_connect_timeout_ms_max;
	int agent_reconnect_delay_ms_min;
	int agent_reconnect_delay_ms_max;
	int agent_heartbeat_timeout_ms_min;
	int agent_heartbeat_timeout_ms_max;
	int agent_heartbeat_interval_ms_min;
	int agent_heartbeat_interval_ms_max;
	int agent_restart_delay_ms_min;
	int agent_restart_delay_ms_max;
} AdbDoctorConf;

typedef struct AdbDoctorConfShm
{
	dsm_segment *seg;
	shm_toc *toc;
	AdbDoctorConf *confInShm;
} AdbDoctorConfShm;

typedef struct AdbDoctorConfRow
{
	char *k;
	char *v;
	bool editable;
	char *comment;
} AdbDoctorConfRow;

static inline void pfreeAdbDoctorConfShm(AdbDoctorConfShm *confShm)
{
	if (confShm)
	{
		dsm_detach(confShm->seg);
		confShm->seg = NULL;
		pfree(confShm);
		confShm = NULL;
	}
}

static inline void pfreeAdbDoctorConfRow(AdbDoctorConfRow *src)
{
	if (src)
	{
		pfree(src->k);
		pfree(src->v);
		pfree(src->comment);
		pfree(src);
		src = NULL;
	}
}

extern void checkAdbDoctorConf(AdbDoctorConf *src);
extern bool equalsAdbDoctorConfIgnoreLock(AdbDoctorConf *conf1,
										  AdbDoctorConf *conf2);
extern AdbDoctorConfShm *setupAdbDoctorConfShm(AdbDoctorConf *conf);
/* after attach, parameter confP host the pointer of conf in shm.  */
extern AdbDoctorConfShm *attachAdbDoctorConfShm(dsm_handle handle,
												char *name);
extern AdbDoctorConf *copyAdbDoctorConfFromShm(AdbDoctorConfShm *shm);
extern void copyAdbDoctorConfAvoidLock(AdbDoctorConf *dest,
									   AdbDoctorConf *src);
extern void refreshAdbDoctorConfInShm(AdbDoctorConf *confInLocal,
									  AdbDoctorConfShm *shm);
extern void validateAdbDoctorConfEditableEntry(char *k, char *v);

extern int selectAdbDoctorConfInt(char *key);
extern char *selectAdbDoctConfByKey(char *key);
extern AdbDoctorConf *selectAllAdbDoctorConf(MemoryContext spiContext);
extern int selectEditableAdbDoctorConf(MemoryContext spiContext,
									   AdbDoctorConfRow **rowDataP);
extern void updateAdbDoctorConf(char *key, char *value);

#endif