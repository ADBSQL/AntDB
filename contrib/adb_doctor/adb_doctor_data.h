/*--------------------------------------------------------------------------
 *
 * Copyright (c) 2018-2019, Asiainfo Database Innovation Lab
 *
 * -------------------------------------------------------------------------
 */
#ifndef ADB_DOCTOR_DATA_H
#define ADB_DOCTOR_DATA_H

#include "storage/dsm.h"
#include "storage/shm_toc.h"
#include "storage/s_lock.h"
#include "postmaster/bgworker.h"
#include "catalog/mgr_node.h"
#include "catalog/mgr_host.h"
#include "mgr/mgr_helper.h"

#define ADB_DOCTOR_SHM_DATA_MAGIC 0x79fb2450

typedef enum Adb_Doctor_Type
{
	ADB_DOCTOR_TYPE_NODE_MONITOR = 1, /* avoid the default value 0 */
	ADB_DOCTOR_TYPE_HOST_MONITOR,
	ADB_DOCTOR_TYPE_SWITCHER,
	ADB_DOCTOR_TYPE_REPAIRER
} Adb_Doctor_Type;

typedef struct AdbDoctorBgworkerData
{
	slock_t mutex;
	/* Use the type and oid as unique identification */
	Adb_Doctor_Type type;
	Oid oid;
	char *displayName;
	/* the handle of a shm that all doctor attached */
	dsm_handle commonShmHandle;
	/* used to judge the doctor is working properly, before use it, 
	 * get it's value from shm. */
	bool ready;
} AdbDoctorBgworkerData;

typedef struct AdbDoctorBgworkerDataShm
{
	dsm_segment *seg;
	shm_toc *toc;
	AdbDoctorBgworkerData *dataInShm;
} AdbDoctorBgworkerDataShm;

typedef struct AdbDoctorBgworkerStatus
{
	dlist_node wi_links;
	BackgroundWorkerHandle *handle;
	pid_t pid;
	BgwHandleStatus status;
	AdbDoctorBgworkerData *bgworkerData;
} AdbDoctorBgworkerStatus;

extern AdbDoctorBgworkerDataShm *
setupAdbDoctorBgworkerDataShm(AdbDoctorBgworkerData *data);
extern AdbDoctorBgworkerData *
attachAdbDoctorBgworkerDataShm(Datum main_arg, char *name);

extern bool isIdenticalDoctorMgrNode(MgrNodeWrapper *data1,
									 MgrNodeWrapper *data2);
extern bool isIdenticalDoctorMgrHost(MgrHostWrapper *data1,
									 MgrHostWrapper *data2);
extern bool isIdenticalDoctorMgrNodes(dlist_head *list1,
									  dlist_head *list2);
extern bool isIdenticalDoctorMgrHosts(dlist_head *list1,
									  dlist_head *list2);

extern void pfreeAdbDoctorBgworkerData(AdbDoctorBgworkerData *src);
extern void pfreeAdbDoctorBgworkerStatus(AdbDoctorBgworkerStatus *src,
										 bool freeData);

extern void logAdbDoctorBgworkerData(AdbDoctorBgworkerData *src,
									 char *title, int elevel);

#endif