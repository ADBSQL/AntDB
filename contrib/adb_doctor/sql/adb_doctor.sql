CREATE EXTENSION adb_doctor;

--
-- These tests don't produce any interesting output.  We're checking that
-- the operations complete without crashing or hanging and that none of their
-- internal sanity tests fail.
--
SELECT adb_doctor_start(2,'');
SELECT adb_doctor_start(1, 60);
SELECT adb_doctor_stop();
