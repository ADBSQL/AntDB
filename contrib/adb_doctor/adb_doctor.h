/*--------------------------------------------------------------------------
 *
 * adb_doctor.h
 *		Definitions for adb doctor
 *
 * Copyright (c) 2018-2019, Asiainfo Database Innovation Lab
 *
 * IDENTIFICATION
 *		contrib/adb_doctor/adb_doctor.h
 *
 * -------------------------------------------------------------------------
 */
#ifndef ADB_DOCTOR_H
#define ADB_DOCTOR_H

#include "postgres.h"
#include "pgstat.h"

#include "storage/procarray.h"
#include "storage/ipc.h"
#include "storage/lwlock.h"
#include "utils/snapmgr.h"
#include "access/xact.h"
#include "adb_doctor_conf.h"
#include "adb_doctor_data.h"

#define ADB_DOCTOR_LAUNCHER_MAGIC 0x79fb2448

#define ADB_DOCTOR_BGW_TYPE_WORKER "antdb doctor worker"
#define ADB_DOCTOR_BGW_TYPE_LAUNCHER "antdb doctor launcher"
#define ADB_DOCTOR_BGW_LIBRARY_NAME "adb_doctor"
#define ADB_DOCTOR_FUNCTION_NAME_LAUNCHER "adbDoctorLauncherMain"
#define ADB_DOCTOR_FUNCTION_NAME_NODE_MONITOR "adbDoctorNodeMonitorMain"
#define ADB_DOCTOR_FUNCTION_NAME_HOST_MONITOR "adbDoctorHostMonitorMain"
#define ADB_DOCTOR_FUNCTION_NAME_SWITCHER "adbDoctorSwitcherMain"
#define ADB_DOCTOR_FUNCTION_NAME_REPAIRER "adbDoctorRepairerMain"

#define ADB_DOCTORS_LAUNCH_OK "OK"
#define ADB_DOCTORS_LAUNCH_FAILURE "FAILURE"

extern void adbDoctorLauncherMain(Datum main_arg) pg_attribute_noreturn();
extern void adbDoctorStopLauncher(bool waitForStopped);
extern void adbDoctorStopBgworkers(bool waitForStopped);
extern void adbDoctorSignalLauncher(void);

extern void adbDoctorNodeMonitorMain(Datum main_arg) pg_attribute_noreturn();
extern void adbDoctorHostMonitorMain(Datum main_arg) pg_attribute_noreturn();
extern void adbDoctorSwitcherMain(Datum main_arg) pg_attribute_noreturn();
extern void adbDoctorRepairerMain(Datum main_arg) pg_attribute_noreturn();

extern void cleanupAdbDoctorBgworker(dsm_segment *seg, Datum arg);
extern void notifyAdbDoctorRegistrant(void);

#endif