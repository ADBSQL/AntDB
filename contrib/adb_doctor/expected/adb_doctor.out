CREATE EXTENSION adb_doctor;
--
-- These tests don't produce any interesting output.  We're checking that
-- the operations complete without crashing or hanging and that none of their
-- internal sanity tests fail.
--
SELECT adb_doctor_start();
 adb_doctor_start 
-------------
 
(1 row)

SELECT adb_doctor_stop();
 adb_doctor_stop 
-----------------------
 
(1 row)

